pennylane_lightning/lightning_kokkos/_state_vector.py

```diff
@@ -26,6 +52,433 @@ class LightningKokkosStateVector:  # pylint: disable=too-few-public-methods
+
+        # get indices for which the state is changed to input state vector elements
+        ravelled_indices = np.ravel_multi_index(unravelled_indices.T, [2] * self._num_wires)
+        return ravelled_indices, state
+
+    def _get_basis_state_index(self, state, wires):
+        """Returns the basis state index of a specified computational basis state.
+
+        Args:
+            state (array[int]): computational basis state of shape ``(wires,)``
+                consisting of 0s and 1s
+            wires (Wires): wires that the provided computational state should be initialized on
+
+        Returns:
+            int: basis state index
+        """
+        # length of basis state parameter
+        n_basis_state = len(state)
+
+        if not set(state.tolist()).issubset({0, 1}):
+            raise ValueError("BasisState parameter must consist of 0 or 1 integers.")
+
+        if n_basis_state != len(wires):
+            raise ValueError("BasisState parameter and wires must be of equal length.")
+
+        # get computational basis state number
+        basis_states = 2 ** (self._num_wires - 1 - np.array(wires))
+        basis_states = qml.math.convert_like(basis_states, state)
+        return int(qml.math.dot(state, basis_states))
+
+    def _apply_state_vector(self, state, device_wires: Wires):
+        """Initialize the internal state vector in a specified state.
+        Args:
+            state (array[complex]): normalized input state of length ``2**len(wires)``
+                or broadcasted state of shape ``(batch_size, 2**len(wires))``
+            device_wires (Wires): wires that get initialized in the state
+        """
+
+        if isinstance(state, self._kokkos_state.__class__):
+            state_data = allocate_aligned_array(state.size, np.dtype(self.dtype), True)
+            state.DeviceToHost(state_data)
+            state = state_data
+
+        ravelled_indices, state = self._preprocess_state_vector(state, device_wires)
+
+        # translate to wire labels used by device
+        output_shape = [2] * self._num_wires
+
+        if len(device_wires) == self._num_wires and Wires(sorted(device_wires)) == device_wires:
+            # Initialize the entire device state with the input state
+            self.sync_h2d(np.reshape(state, output_shape))
+            return
+
+        self._kokkos_state.setStateVector(ravelled_indices, state)  # this operation on device
+
+    def _apply_basis_state(self, state, wires):
+        """Initialize the state vector in a specified computational basis state.
+
+        Args:
+            state (array[int]): computational basis state of shape ``(wires,)``
+                consisting of 0s and 1s.
+            wires (Wires): wires that the provided computational state should be
+                initialized on
+
+        Note: This function does not support broadcasted inputs yet.
+        """
+        num = self._get_basis_state_index(state, wires)
```

Comment:
Maybe less of a case than `_create_basis_state` since `_get_basis_state_index` has some logic which could be reuse elsewhere.

Reply:
You are right. I will merge the function `_create_basis_state` with `_apply_basis_state`. However, It would be better to keep  `_get_basis_state_index`  for readability.